Use buffer language when formatting with Prettier

JPGibb opened the pull request:
Set `prettier_parser` explicitly if the file extension for the buffer does not match a known one for the current language

Release Notes:

- N/A

JPGibb committed:
Use the buffer language when formatting with prettier if the file extension does not match known file extensions for the buffer.

SomeoneToIgnore reviewed (commented):
Can you please provide more context on what's broken here, how to test the fix, etc.?

JPGibb commented:
Hi,

In my case I have json that gets stored in files like "feature.settings".
When editing a buffer without an associated file formatting works correctly as the prettier parser assiociated with the language is used.
However, after saving this buffer to a file like the on above, formatting would fail with the error:

**UndefinedParserError: No parser could be inferred for file {path_to_file}**

So for a file like 'feature.settings':
``` json
{
  "foo": "foo",
"bar": "misformatted"
}
```

Formatting would fail, even with the buffer language in Zed set to json.

With this change, the prettier parser would be set explicitly using the buffer's language rather than inferred from the file extension.
So the file would be correctly formatted as json.

Cheers

SomeoneToIgnore reviewed (commented):
Thank you for the context: sounds like this is quite a logic-related one, and we'd better have a test so it does not degrade later.

We have `async fn test_document_format_with_prettier` and other prettier tests, let's write another one for this scenario.

JPGibb committed:
Add test for prettier formatting with language set explicitly

JPGibb commented:
Hi, Sorry for the delay getting back to you. Long week :).

I've added a new test `test_document_format_with_prettier_explicit_language` which passes.
However, I noticed after doing this that the code I wrote is in a Self::Real block whereas the test uses Self::Test.

I'm not entirely sure how to handle this, is there some way I can set my test up to run the code that I changed?

Thanks

SomeoneToIgnore reviewed (commented):
Thank you for the test, and no worries, I have other things to do anyway.

What if we adjust 
`let formatted_text = buffer.text() + FORMAT_SUFFIX;` to write the parser name too?

That will require amending a few existing tests and restructuring the code of `async fn format` so that the common code for parser determination is shared between both test and real branches.

But after that, the code before the change should output a totally different value, which we can assert.

JPGibb committed:
Add parser language to end of TestPrettier output

JPGibb committed:
Move Parser determination logic into a closure used by real and test prettier

SomeoneToIgnore reviewed (approved):
Looking at the CI and the changes, this turned out to be the first test that uses Prettier parsers...

Thank you a lot for adding one.
I've left a few minor style nits and will be happy to merge this after today's release, just in case.

SomeoneToIgnore committed:
Tidy up
